enumerables.rb

```diff
@@ -0,0 +1,184 @@
+      i = 0
+      new_arr = []
+      new_hash = {}
+      while i < length
+        if instance_of?(Array)
+          new_arr << self[i] if yield self[i]
+        elsif yield(keys[i], values[i])
+          new_hash[keys[i]] = values[i]
+        end
+        i += 1
+      end
+
+      return new_arr if instance_of?(Array)
+      return new_hash if instance_of?(Hash)
+
+    else
+      puts 'block was not given'
+    end
+  end
+
+  def my_all?(sth = nil)
+    return true if empty?
+
+    if block_given?
+      my_each do |item|
+        return false unless yield item
+      end
+    else
+      my_each do |item|
+        if sth.instance_of?(Regexp)
+          return false if item.match?(sth) == false
+        elsif sth.is_a? Class
+          return false unless item.is_a? sth
+        elsif [false, nil].include?(item)
+          return false
+        end
+      end
+    end
+    true
+  end
+
+  def my_any?(sth = nil)
+    return false if empty?
+
+    if block_given?
+      my_each do |item|
+        return true if yield item
+      end
+    else
+      my_each do |item|
+        if sth.instance_of?(Regexp)
+          return true if item.match?(sth)
+        elsif sth.is_a? Class
+          return true if item.is_a? sth
+        else
+          return false if self == [nil] || self == [false] || self == [nil, false] || self == [false, nil]
+
+          return true
+        end
+      end
+    end
+    false
+  end
+
+  def my_none?(sth = nil)
+    return true if empty?
+
+    if block_given?
+      my_each do |item|
+        return false if yield item
+      end
+    else
+      my_each do |item|
+        if sth.instance_of?(Regexp)
+          return false if item.match?(sth)
+        elsif sth.is_a? Class
+          return false if item.is_a? sth
+        else
+          return true if self == [nil] || self == [false] || self == [nil, false] || self == [false, nil]
+
+          return false
+        end
+      end
+    end
+    true
+  end
+
+  def my_count(arg = nil)
+    i = 0
+    if arg
+      my_each do |item|
+        i += 1 if item == arg
+      end
+
+    elsif block_given?
+      my_each do |item|
+        i += 1 if yield item
+      end
+
+    else
+      i = to_a.length
+    end
+    i
+  end
+
+  def my_map
+    result = []
+    if proc
+      to_a.my_each do |item|
+        result.push(proc.call(item))
+      end
+    else
+      to_a.my_each do |item|
+        result.push(yield item)
+      end
+    end
+    result
+  end
+
+  def my_inject(*sth)
+    new_arr = self
+    if block_given?
+      now_it_is = sth[0] if sth.length == 1
+      now_it_is = new_arr.shift unless sth.nil? || sth.length == 1
+      new_arr.my_each do |item|
+        now_it_is = yield(now_it_is, item)
+      end
+    else
+      now_it_is = sth.length > 1 ? sth.shift : new_arr.shift
+      new_arr.my_each do |item|
+        now_it_is = now_it_is.send(sth[0].to_s, item)
+      end
+      return now_it_is
+    end
+    now_it_is
+  end
+end
+
```

Comment:
- [ ] enumerables #my_inject raises a "LocalJumpError" when no block or argument is given
      Failure/Error: expect { array.my_inject }.to raise_error(LocalJumpError)

- [ ]  enumerables #my_inject does not mutate the original array

- [ ] enumerables #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
      Failure/Error: expect(array.my_inject(&operation)).to eq array.inject(&operation)

- [ ]  enumerables #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range
      Failure/Error: actual = range.my_inject { |prod, n| prod * n }
      
      NoMethodError:
        undefined method `shift' for 5..50:Range

- [ ] enumerables #my_inject when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
      Failure/Error: actual = range.my_inject(4) { |prod, n| prod * n }
      
      NameError:
        undefined local variable or method `length' for 5..50:Range

- [ ]  enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
      Failure/Error: expect(range.my_inject(:*)).to eq range.inject(:*)
      
      NoMethodError:
        undefined method `shift' for 5..50:Range

- [ ]  enumerables #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range
      Failure/Error: expect(range.my_inject(2, :*)).to eq range.inject(2, :*)
      
      NameError:
        undefined local variable or method `length' for 5..50:Range